components/eip681/src/parse.rs

```diff
@@ -0,0 +1,1565 @@
+//! Types and functions used for parsing.
+
+use std::borrow::Cow;
+
+use nom::{AsChar, Parser};
+use snafu::{OptionExt, ResultExt};
+
+use crate::error::*;
+
+/// Zero or more consecutive digits.
+///
+/// ```abnf
+/// *DIGIT
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct Digits {
+    places: Vec<u8>,
+}
+
+impl core::fmt::Display for Digits {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for n in self.places.iter() {
+            f.write_fmt(format_args!("{n}"))?;
+        }
+        Ok(())
+    }
+}
+
+impl Digits {
+    #[cfg(any(test, feature = "test-dependencies"))]
+    pub const MAX_PLACES: usize = u64::MAX.ilog10() as usize;
+
+    /// Returns the `u64` representation.
+    ///
+    /// ## Errors
+    /// Errors if internal arithmetic operations overflow.
+    pub fn as_u64(&self) -> Result<u64, ValidationError> {
+        let mut total = 0u64;
+        for (mag, digit) in self.places.iter().rev().enumerate() {
+            let increment = (*digit as u64)
+                .checked_mul(
+                    10u64
+                        .checked_pow(u32::try_from(mag).context(IntegerSnafu)?)
+                        .context(OverflowSnafu)?,
+                )
+                .context(OverflowSnafu)?;
+            total = total.checked_add(increment).context(OverflowSnafu)?;
+        }
+        Ok(total)
+    }
+
+    /// Returns the digits as a ratio, where prefixed zeros
+    /// are treated as a denominator.
+    fn as_decimal_ratio(&self) -> Result<(u64, u64), ValidationError> {
+        let denominator = 10u64
+            .checked_pow(u32::try_from(self.places.len()).context(IntegerSnafu)?)
+            .context(OverflowSnafu)?;
+        let mut numerator = 0u64;
+
+        for digit in self.places.iter() {
+            numerator = numerator
+                .checked_mul(10)
+                .context(OverflowSnafu)?
+                .checked_add(*digit as u64)
+                .context(OverflowSnafu)?;
+        }
```

Comment:
And then it's clear these are identical:
```suggestion
        let numerator = self.as_u64();
```